contracts/zaps/ZapTrancheLp.sol

```diff
@@ -0,0 +1,277 @@
+    ///                        the data should be encoded with abi.encode(data, "PermitData[]")
+    ///                        each PermitData struct provided will be executed as a call.
+    ///                        An example use of this is if using a token with permit like USDC
+    ///                        to encode a permit which gives this contract allowance before minting.
+    function hopToTranche(
+        uint256 _toMint,
+        Out memory _ptOutInfo,
+        Out memory _ytOutInfo,
+        In memory _ptInInfo,
+        In memory _ytInInfo,
+        bool _onlyPrincipal,
+        PermitData[] calldata _permitCallData
+    ) public notFrozen {
+        IERC20 ytPool;
+        // get tranches
+        ITrancheExt trancheFrom = ITrancheExt(
+            IPool(address(uint160(uint256(_ptOutInfo.poolId)) >> (12 * 8)))
+                .bond()
+        );
+
+        ITrancheExt trancheTo = ITrancheExt(
+            IPool(address(uint160(uint256(_ptInInfo.poolId)) >> (12 * 8)))
+                .bond()
+        );
+
+        // transfer pt lp tokens
+        IERC20 ptPool = IERC20(
+            address(uint160(uint256(_ptOutInfo.poolId)) >> (12 * 8))
+        );
+        ptPool.transferFrom(
+            msg.sender,
+            address(this),
+            ptPool.balanceOf(msg.sender)
+        );
+
+        // exit PT position
+        _vault.exitPool(
+            _ptOutInfo.poolId,
+            address(this),
+            payable(address(this)),
+            _ptOutInfo.request
+        );
+
+        // withdraw principal here. This should accumulate underlying.
+        trancheFrom.withdrawPrincipal(
+            trancheFrom.balanceOf(address(this)),
+            address(this)
+        );
+
+        if (!_onlyPrincipal) {
+            // transfer yt lp tokens
+            ytPool = IERC20(
+                address(uint160(uint256(_ytOutInfo.poolId)) >> (12 * 8))
+            );
+            ytPool.transferFrom(
+                msg.sender,
+                address(this),
+                ytPool.balanceOf(msg.sender)
+            );
+
+            // exit YT position
+            _vault.exitPool(
+                _ytOutInfo.poolId,
+                address(this),
+                payable(address(this)),
+                _ytOutInfo.request
+            );
+
+            // withdraw interest to the new tranche to prep a pre-funded deposit
+            trancheFrom.withdrawInterest(
+                IERC20(address(trancheFrom.interestToken())).balanceOf(
+                    address(this)
+                ),
+                address(this)
+            );
+        }
+
+        // transfer required amount and run prefunded deposit
+        trancheFrom.underlying().transfer(
+            address(trancheTo.position()),
+            _toMint
+        );
+        trancheTo.prefundedDeposit(address(this));
+
+        uint256 ptLpBalanceBefore = ptPool.balanceOf(msg.sender);
```

Comment:
Instead of querying before and checking for increase we could just check for total balance after and let the front end calculate the diff if there's some before.